packages/store-api/src/platforms/vtex/resolvers/validateCart.ts

```diff
@@ -0,0 +1,162 @@
+    orderFormId,
+    messages,
+    salesChannel,
+    items: items.map(
+      ({ uniqueId, quantity, seller, sellingPrice, availability }) => ({
+        uniqueId,
+        quantity,
+        seller,
+        sellingPrice,
+        availability,
+      })
+    ),
+  })
+
+  return JSON.stringify(pick(of1)) === JSON.stringify(pick(of2))
+}
+
+/**
+ * This resolver implements the optimistic cart behavior. The main idea in here
+ * is that we receive a cart from the UI (as query params) and we validate it with
+ * the commerce platform. If the cart is valid, we return null, if the cart is
+ * invalid according to the commerce platform, we return the new cart the UI should use
+ * instead
+ *
+ * The algoritm is something like:
+ * 1. Fetch orderForm from VTEX
+ * 2. Compute delta changes between the orderForm and the UI's cart
+ * 3. Update the orderForm accordingly
+ * 4. If any chages were made, send to the UI the new cart. Null otherwise
+ */
+export const validateCart = async (
+  _: unknown,
+  {
+    cart: {
+      order: { orderNumber, acceptedOffer },
+    },
+  }: { cart: IStoreCart },
+  ctx: Context
+) => {
+  const {
+    clients: { commerce },
+    loaders: { skuLoader },
+  } = ctx
+
+  // Step1: Get OrderForm from VTEX Commerce
+  const orderForm = await commerce.checkout.orderForm({
+    id: orderNumber,
+  })
+
+  // Step2: Process items from both browser and checkout so they have the same shape
+  const browserItemsById = groupById(acceptedOffer)
+  const originItemsById = groupById(orderForm.items.map(orderFormItemToOffer))
+  const browserItems = Array.from(browserItemsById.values()) // items on the user's browser
+  const originItems = Array.from(originItemsById.values()) // items on the VTEX platform backend
+
+  // Step3: Compute delta changes
+  const { itemsToAdd, itemsToUpdate } = browserItems.reduce(
+    (acc, item) => {
+      const maybeOriginItem = originItemsById.get(getId(item))
+
+      if (!maybeOriginItem) {
+        acc.itemsToAdd.push(item)
+      } else {
+        acc.itemsToUpdate.push({
+          ...maybeOriginItem,
+          quantity: item.quantity,
+        })
+      }
+
+      return acc
+    },
+    {
+      itemsToAdd: [] as IStoreOffer[],
+      itemsToUpdate: [] as IStoreOffer[],
+    }
+  )
+
+  const itemsToDelete = originItems
+    .filter((item) => !browserItemsById.has(getId(item)))
+    .map((item) => ({ ...item, quantity: 0 }))
+
+  const changes = [...itemsToAdd, ...itemsToUpdate, ...itemsToDelete].map(
+    offerToOrderItemInput
+  )
+
+  if (changes.length === 0) {
+    return null
+  }
+
+  // Step4: Apply delta changes to order form
+  const updatedOrderForm = await commerce.checkout.updateOrderFormItems({
+    id: orderForm.orderFormId,
+    orderItems: changes,
+  })
+
+  // Step5: If no changes detected before/after updating orderForm, the order is validated
+  if (equals(orderForm, updatedOrderForm)) {
+    return null
```

Comment:
With the change I proposed a few lines above, this condition would just return the same orderForm to the client, but with the `updated` property set to `false`.